### PR TITLE
Fix fetch implementation with CF workers

### PR DESCRIPTION
**Pull request by SferaDev**
Closes: #59 

In #58 we plan to use the API methods internally and stop using ``fetch`` directly (which also solves the `Illegal Invocation` issue). This is only a fix so we can bump a ``patch`` version early and solve the compatibility with CF.

### CODE REVIEW

File: client/src/index.ts [hunk 1]
```diff
@@ -403,8 +396,9 @@ export class RestRepository<T> extends Repository<T> {
   async request<T>(method: string, path: string, body?: unknown): Promise<T | undefined> {
     const { databaseURL, apiKey } = this.client.options;
     const branch = await this.client.getBranch();
+    const fetchImpl = this.fetch;
```
Comment (SferaDev) [inline client/src/index.ts:399]: This is actually the fix 🤯

File: client/src/index.ts [hunk 2]
```diff
@@ -384,16 +384,9 @@ export class RestRepository<T> extends Repository<T> {
...
-
-    if (doWeHaveFetch) {
-      this.fetch = fetch;
-    } else if (isInjectedFetchProblematic) {
-      throw new Error(errors.falsyFetchImplementation);
-    } else {
-      this.fetch = this.client.options.fetch;
-    }
+    const fetchImpl = typeof fetch !== 'undefined' ? fetch : this.client.options.fetch;
+    if (!fetchImpl) throw new Error('No fetch implementation provided');
```
Comment (gimenete) [inline client/src/index.ts:388]: I see some value in this check:

```ts
    if (!fetchImpl && 'fetch' in this.client.options) throw new Error('Invalid fetch implementation provided');
```

That was the idea with the "falsy" error message.

In any case we should remove the `errors.falsyFetchImplementation` in the errors file if we don't use it anymore.

File: client/src/index.ts (hunk 1 above)
Comment (gimenete) [reply client/src/index.ts:399]: Wow! Great find!

Yeah, no need to bind IMO.

**Review by gimenete — APPROVED** [review]
Thanks for the fix!

**Comment by ncheikh** [comment]
Thanks for the fix! Tested with this branch and my CF workers now behave :slightly_smiling_face:

File: client/src/index.ts (hunk 1 above)
Comment (ncheikh) [reply client/src/index.ts:399]: @SferaDev Could you please explain why this works vs just using this.fetch in the call?

File: client/src/index.ts (hunk 2 above)
Comment (SferaDev) [reply client/src/index.ts:388]: Since it's not related to the fix on CF, let's move the discussion to #56

File: client/src/index.ts (hunk 1 above)
Comment (SferaDev) [reply client/src/index.ts:406]: @ncheikh It's not a simple explanation 😢 

If you're curious I strongly suggest you to read [YDKJS 1st edition](https://github.com/getify/You-Dont-Know-JS/blob/1st-ed/this%20&%20object%20prototypes/README.md#you-dont-know-js-this--object-prototypes). 

What is weird is that it does work on node.js but not on V8 (CF Workers JS runtime), I would have to dig deeper to know the exact cause.

Comment (ncheikh) [reply client/src/index.ts:406]: Thank you @SferaDev, I will take a look!  I was a bit baffled why it wasn't working and tried several things but it wasn't obvious to me why it wasn't working for CF as it was working in other environments/engines. I wanted to open an MR for the fix but couldn't come up with a solution and was hopeful y'all would be able to sort it out and add clarity.

I very much appreciate the quick turn around and taking the time to answer my question!

**Pull request merged**